rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/bpp/LogMethodsByRequiredAnnotationsBeanPostProcessor.kt

```diff
@@ -0,0 +1,92 @@
+package systems.ajax.englishstudytelegrambot.bpp
+
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.cglib.proxy.Enhancer
+import org.springframework.cglib.proxy.MethodProxy
+import org.springframework.stereotype.Component
+import systems.ajax.englishstudytelegrambot.annotation.LogMethodsByRequiredAnnotations
+import java.lang.reflect.Method
+import kotlin.reflect.KClass
+import org.springframework.cglib.proxy.MethodInterceptor;
+
+
+@Component
+class LogMethodsByRequiredAnnotationsBeanPostProcessor : BeanPostProcessor {
+
+    private val mapBeanNameBeanClassAndAnnotatedMethods: MutableMap<String, Pair<Class<*>, List<Method>>> =
+        mutableMapOf()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any {
+        val beanClass = bean.javaClass
+        if (beanClass.isAnnotationPresent(LogMethodsByRequiredAnnotations::class.java)) {
+            mapBeanNameBeanClassAndAnnotatedMethods[beanName] =
+                beanClass to findMethodsWithRequiredAnnotations(beanClass)
+        }
+        return bean
+    }
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any =
+        mapBeanNameBeanClassAndAnnotatedMethods.get(beanName)?.let { enhanceBean(bean, beanName) } ?: bean
+
+    private fun findMethodsWithRequiredAnnotations(beanClass: Class<*>): List<Method> {
+        val requiredAnnotations: Array<out KClass<out Any>> =
+            beanClass.getAnnotation(LogMethodsByRequiredAnnotations::class.java).requiredAnnotations
+        return beanClass
+            .getDeclaredMethods()
+            .asSequence()
+            .filter { method ->
+                method.annotations
+                    .map { it.annotationClass }
+                    .any { requiredAnnotations.contains(it) }
+            }
+            .toList()
+    }
+
+    private fun enhanceBean(bean: Any, beanName: String): Any {
+        val enhancer = Enhancer()
+        enhancer.setSuperclass(mapBeanNameBeanClassAndAnnotatedMethods[beanName]!!.first)
```

Comment:
Map values should be accessed safely. Using the non-null assertion operator here can throw a NullPointerException.